### ISSUE TITLE
Reproducing CIFAR results

### ISSUE THREAD

**Issue by MacroMayhem** [teddykoker/torchsort#19]
Thanks a lot for this implementation. 
I was wondering how can I use the repo to reproduce the results on CIFAR as reported in the paper. As I understand, the target one-hot encoding will serve as top-k classification(k=1). But, after obtaining the logits and passing through the softmax(putting output [0, 1])  the objective is to make the output follow the target ordering. How can this be achieved?

**Comment by teddykoker** [comment]
In the paper they follow the method described in [Differentiable Ranking and Sorting using Optimal Transport
](https://papers.nips.cc/paper/2019/hash/d8c24ca8f23c562a5600876ca2a550ce-Abstract.html):

<img width="544" alt="Screen Shot 2021-05-03 at 4 16 57 PM" src="https://user-images.githubusercontent.com/11153048/116928465-1c1a7800-ac2b-11eb-9571-81bce2f00945.png">
<img width="542" alt="Screen Shot 2021-05-03 at 4 17 33 PM" src="https://user-images.githubusercontent.com/11153048/116928474-1e7cd200-ac2b-11eb-940e-66b324185956.png">

I believe this is what you would need to do:

```python
import torch
import torchsort

def J(u, k=1):
    return torch.nn.functional.relu(u - k + 1)

BATCH_SIZE = 64
L = 10 # number of classes

f_w = torch.randn(BATCH_SIZE, L) # logits of the CNN
l = torch.randint(high=L, size=(BATCH_SIZE, 1)) # integer labels

R = torchsort.soft_rank(f_w) # compute ranks of logits

R_l = ranks.gather(-1, l) # index ranks at label (l)

loss = J(L - R_l).mean()
loss.backward()
...
```

Intuitively this makes sense as we are effectively maximizing the rank of the correctly labelled logit; however I am not 100% positive this is correct. Once I have some more time on my hands I will work to fully reproduce their results.

**Comment by teddykoker** [comment]
Alright training on CIFAR10 was quite straightforward, I uploaded an example to [extra/cifar10.py](https://github.com/teddykoker/torchsort/blob/main/extra/cifar10.py), which seems to performing about similar to the graph in the paper. Neither this paper nor the previously mentioned one give exact details about the architecture they use (kernel size, hidden size, etc.) so I think this is probably the best you can do in terms of reproduction without reaching out to the original authors for these details.

**Comment by MacroMayhem** [comment]
Wow! this was super helpful. Thanks @teddykoker

**Comment by vltanh** [comment]
Hi, thanks for the amazing work. However, I tried to compare this to the conventional cross-entropy loss and notice a few things:
- Augmentation is extremely important (without it, it can only lead to 75% accuracy). The paper didn't go into details or provide a reproducible implementation, so I guess that's that.
- Cross entropy loss performs better, which does not match with the claim in the paper. I wonder if this is the issue with the current implementation of the loss or this approach is simply just not as good as the paper claimed. 
 
I found the implementation for the loss in Cuturi et al., 2019 [here](https://github.com/google-research/google-research/blob/master/soft_sort/losses.py). Would you mind checking it out? 

The paper also mentions that
> Similarly to Cuturi et al. (2019), we found that squashing the scores θ to [0, 1]^n with a logistic map was beneficial.
I wonder what this means. Should we do a softmax before ranking? Does it matter?

**Comment by teddykoker** [comment]
Thanks for the comment @vltanh. Following you advice I have made a few changes to the cifar10 script (still a work in progress on the [fix_topk](https://github.com/teddykoker/torchsort/tree/fix_topk) branch). Adding a softmax prior to the topk loss yields a test accuracy about the same as cross entropy:

![](https://raw.githubusercontent.com/teddykoker/torchsort/fix_topk/extra/cifar10_test_accuracy.png)

Since these are both considerably higher than the paper reports, I will be trying again without data augmentation. Additionally the regularization coefficient may need to be tuned to obtain the best results.

**Comment by vltanh** [comment]
I have done experiments with both not using softmax and using softmax and I find that they achieve similar results (still, it's just on 1 test, so I'm not sure about that either).

For the current case where you use augmentation, my result is similar to you: CELoss outperforms at almost every step until the end where the two performs similarly. 

For the case where you don't use augmentation, you might see that: CELoss quickly overfits (accuracy drops after the first 30-50 epochs) while TopK loss gradually increases, but they reach similar accuracy at the end too (on CIFAR100, TopK is worse).

About tuning the regularization coefficient, I tried with L2 for 0.1, 1.0, 10.0 and they achieve similar results, although 1.0 is marginally better.

**Comment by teddykoker** [comment]
Here is what I got with no augmentation; this looks similar to what you mention:

![cifar10_test_accuracy](https://user-images.githubusercontent.com/11153048/125295025-cfba7b00-e2f2-11eb-9647-37066738a525.png)

**Comment by teddykoker** [comment]
I think at this point it would make sense to reach out to the original authors and see if they could provide any insight as to what we are doing wrong. My soft sort/rank is tested to perform numerically identically to the original implementation, so it is likely just an issue with the loss function, or some difference in CNN architecture that provides different results.
